enhancements/installer/byo-custom-dns.md

```diff
@@ -0,0 +1,571 @@
+enhancement proposal will not talk about how these entries are configured for
+the OpenShift Managed offering. 
+
+### User Stories
+
+"As a cloud admin, I want to run an OpenShift cluster on AWS with my custom DNS
+solution in place of the default cloud DNS."
+
+TODO: Is there anything else to add here?
+
+### Goals
+
+- Enable OpenShift customers to use their custom DNS solution instead of the
+default cloud solution.
+- Start with AWS and lay the groundwork to add this capability for other cloud
+platforms like Azure and GCP.
+- Continue using the default cloud LB service for API, API-Int and Ingress.
+
+### Non-Goals
+
+We will not use the [Kubernetes External DNS provider](https://github.com/openshift/external-dns)
+and/or the [OpenShift External DNS Operator](https://github.com/openshift/external-dns-operator)
+to program dns entries across varied DNS providers. The External DNS Operator
+relies on being able to program the customer's DNS solution. The customer DNS
+solutions in consideration here do not allow OpenShift to configure them.
+
+## Proposal
+
+For the OpenShift installation to make use of a custom DNS solution that is not
+configurable by OpenShift, the customer is responsible for configuring the DNS
+entries for the API, API-Int and Ingress services.
+
+Currently, during cluster installation, OpenShift is repsonsible for adding
+entries for API, API-Int and *.apps into the cloud DNS. These URLs are
+configured with the LB DNS Names of the LBs associated with each of these
+services. Hence the LBs have to be created before they can used to configure
+the DNS solution. Also, important to note is that these LB DNS Names cannot
+be generated before they are created.
+
+When it is indicated to the OpenShift installer via an input parameter that the
+custom DNS solution is in use, OpenShift does not configure the cloud DNS
+solution [Route53 for AWS](https://aws.amazon.com/route53/) for any of its
+services. In the case of API and API-Int services, the customer is responsible
+for creating the external and internal LBs [Elastic Load Balancer](https://aws.amazon.com/elasticloadbalancing/)
+prior to cluster installation. Their custom DNS solution also needs to be
+configured with entries for API and API-Int before cluster installation. OpenShift
+will continue to configure and manage these cloud LBs created by the user.
+
+DNS configuration for Ingress services will be handled differently. The Ingress
+Operator utilizes the upstream k8s LoadBalancer service to create and configure
+the LBs for the *.apps service. In this case, the customer pre-creating the LB
+would mean the LB service would have to be modified to consume a pre-created LB.
+Accomplishing this would require a change to the upstream LB service potentially
+adding delays towards the completion of this feature. Since the LB DNS Name for
+the *.apps would not be available before cluster installation, the DNS solution
```

Comment:
> Does this mean, even with the workaround, the customer cannot have their DNS solution pre-configured? 
That is correct.
>(minus the LB). Why is that?
To pre-configure the customer's DNS solution for *.apps, they would have to know the LB DNS name for the Ingress LB. That will not be available unless the LB service is started by the cluster-ingress-operator. 
We could potentially ask the customer to manually start the LB service (we could help them by providing them with a manifest that they can use) before cluster install and use the Ingress LB DNS name in the Status field of this service to pre-configure the customer DNS solution. We are not pursuing this course at this time.

Reply:
> manually start the LB service (we could help them by providing them with a manifest that they can use) before cluster install and use the Ingress LB DNS name in the Status field of this service to pre-configure the customer DNS solution

If the load balancer service can use an existing LB, we should do that, right? I think that would be preferable to running coredns in the cluster. But is it possible to use a pre-existing LB in this case? I thought not, but hope I am wrong.

Reply:
Starting the Ingress Load Balancer service will create the Load Balancer for you. So, if we want to avoid running the CoreDNS pod, we would have to ask the user to create the Load Balancer service with the same annotations that the cluster-ingress-operator would add. This will cause the Ingress Load Balancer to be created and [`LoadBalancerIngress`](https://github.com/kubernetes/api/blob/cdff1d4efea5d7ddc52c4085f82748c5f3e5cc8e/core/v1/types.go#L4554) to be updated with either the Ingress Load Balancer's IP or Hostname depending on the underlying platform.
The user can then add their *.apps entry to their custom DNS pointing to either the Hostname or IP address of their Ingress Load Balancer. When the cluster-ingress-operator is started within the cluster, it doesn't start a new LB service and continues to reconcile on the existing one. That is why it is important for the user to create the Ingress LB service exactly the way the ingress-operator would. This will work but has the following disadvantages:
1. When the API and/or annotations on the Load Balancer service API changes, the changes within cluster-ingress-operator and the instructions provided to the user to pre-create this service have to be in lock step.
2. Seems like the number of user's pre-create, pre-configure steps are too many thus increasing the chances of configuration errors. (why don't we ask the user to install by UPI instead?)

Reply:
Great! This seems like a promising option to me. I assume the installer rather than the customer could apply the annotations?

Reply:
@patrickdillon you really don't like running a CoreDNS pod! :-) I wanted to use some version of this approach for ARO. See this https://github.com/openshift/enhancements/pull/1400#discussion_r1220167296. But, we can figure out platform specific annotations for AWS and add them.

Ok. Let us figure out what the Installer would do vs the user. Is this what is being suggested?
1. User starts the Ingress Load Balancer service that is generic i.e does not have platform specific/release specific  annotations.
2. Installer has to detect that this LB service is running and add these platform specific annotations (Afaik, Installer cannot do that before the Ingress operator is started in the cluster. This is the part that I am not clear about.)
3. Ingress operator continues to reconcile on this running Load Balancer service and manage it for the life of the cluster.

Reply:
> @patrickdillon you really don't like running a CoreDNS pod! :-)

Guilty! 

> Ok. Let us figure out what the Installer would do vs the user. Is this what is being suggested?

I was thinking more along the lines of:

1. User creates empty load balancer, same as api-int and api, if it's being used
2. User specifies load balancer name in install config
3. Installer retrieves LB name (using aws api) 
4. Installer adds any necessary annotations/labels/metadata to LB
5. Installer lays down manifest(s) for the ingress service in the default-router namespace, populating the ingress hostname with the LB name from step 3

@bennerv pointed us to the manifests ARO uses for this pattern, so they could serve as a guide: https://github.com/jewzaam/installer-aro/commit/80c2ca1a9fddcba68d9251e6ab4f0e68097f509d

Reply:
There is discussion of the ARO implementation in the "Variations" section of the doc.  
As soon as the Load Balancer service is started, it will create an ELB. I am looking into what annotations could be useful for a solution that resembles these steps.
The Installer creates the manifest for the [Ingress controller ](https://github.com/openshift/api/blob/master/operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml) but the annotations need to be added to the LoadBalancer service started by the cluster-ingress-operator. It would be ideal if the cluster-ingress-operator added the annotations to the LB service. Any information to be passed to the ingress-operator from the Installer for the annotations can happen via the IngressController CRD.

Reply:
> As soon as the Load Balancer service is started, it will create an ELB.

Right. I think the cloud provider will reconcile an ELB for the service. So the crucial question is whether it is possible when creating a load balancer service for the cloud provider to accept an existing LB (and not create a new one)? Can we do this by creating a LB that matches the specs in the service so that the cloud provider's reconciliation loop indicates that the LB already exists and does not need to be created?

> I am looking into what annotations could be useful for a solution **that resembles these steps.**

(emphasis mine) Yeah, I don't like in the steps I suggested that the installer is laying down the service. I should have kept that more open-ended, it would be ideal if the cluster ingress operator still laid it down, the installer could pass whatever the needed information is in the API. 

> *I am looking into what annotations could be useful* for a solution that resembles these steps.

It seems like this idea has passed basic sanity tests, so it would be a good idea to get input from the network edge team on this.

Reply:
I want to make sure I have figured out the path forward for AWS, Azure and GCP before I pull in the network edge team. The current proposed solution (with CoreDNS pod) works for all 3 platforms and so the alternative with annotations should also work for all the cloud platforms in consideration. The last time I checked, there was not a consistent story between the 3 platforms.

Reply:
"Variations" has been updated with information about usable annotations in AWS, Azure and GCP.  The network edge team has provided their input. So, marking this discussion as resolved.